server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OnHeapStarTreeBuilder.java

```diff
@@ -66,34 +62,136 @@ public List<StarTreeDocument> getStarTreeDocuments() {
+            // TODO : if all the dimensions are null, then we can skip adding the document
+            // TODO : if we come out of this loop with the very first document dimensions coming with doc iterators exhausted,
+            // then we need to return empty star tree
+            // TODO : we can save empty iterator for dimensions which are not part of segment
+            starTreeDocuments[currentDocId] = getSegmentStarTreeDocument(currentDocId, dimensionReaders, metricReaders);
         }
-
         return sortAndAggregateStarTreeDocuments(starTreeDocuments);
     }
 
+    @Override
+    public void build(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        build(mergeStarTrees(starTreeValuesSubs));
+    }
+
+    /**
+     * Sorts and aggregates the star-tree documents from multiple segments and builds star tree based on the newly
+     * aggregated star-tree documents
+     *
+     * @param starTreeValuesSubs StarTreeValues from multiple segments
+     * @return iterator of star tree documents
+     */
+    Iterator<StarTreeDocument> mergeStarTrees(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        return sortAndAggregateStarTreeDocuments(mergeStarTreeValues(starTreeValuesSubs), true);
+    }
+
+    /**
+     * Returns an array of all the starTreeDocuments from all the segments
+     *
+     * @param starTreeValuesSubs StarTreeValues from multiple segments
+     * @return array of star tree documents
+     */
+    StarTreeDocument[] mergeStarTreeValues(List<StarTreeValues> starTreeValuesSubs) throws IOException {
+        List<StarTreeDocument> starTreeDocuments = new ArrayList<>();
+        for (StarTreeValues starTreeValues : starTreeValuesSubs) {
+            List<Dimension> dimensionsSplitOrder = starTreeValues.getStarTreeField().getDimensionsOrder();
+            SequentialDocValuesIterator[] dimensionReaders = new SequentialDocValuesIterator[starTreeValues.getStarTreeField()
+                .getDimensionsOrder()
+                .size()];
+
+            for (int i = 0; i < dimensionsSplitOrder.size(); i++) {
+                String dimension = dimensionsSplitOrder.get(i).getField();
+                dimensionReaders[i] = new SequentialDocValuesIterator(starTreeValues.getDimensionDocValuesIteratorMap().get(dimension));
+            }
+
+            List<SequentialDocValuesIterator> metricReaders = new ArrayList<>();
+            for (Map.Entry<String, DocIdSetIterator> metricDocValuesEntry : starTreeValues.getMetricDocValuesIteratorMap().entrySet()) {
+                metricReaders.add(new SequentialDocValuesIterator(metricDocValuesEntry.getValue()));
+            }
+
+            boolean endOfDoc = false;
+            int currentDocId = 0;
+            int numSegmentDocs = Integer.parseInt(
+                starTreeValues.getAttributes().getOrDefault(NUM_SEGMENT_DOCS, String.valueOf(DocIdSetIterator.NO_MORE_DOCS))
```

Comment:
I don't see this being set anywhere

Reply:
This will be set in file formats when we read the meta file.

Reply:
+1